exchanges/coinbasepro/coinbasepro_websocket.go

```diff
@@ -46,423 +77,578 @@ func (c *CoinbasePro) WsConnect() error {
+func statusToStandardStatus(stat string) (order.Status, error) {
+	switch stat {
+	case "PENDING":
+		return order.New, nil
+	case "OPEN":
+		return order.Active, nil
+	case "FILLED":
+		return order.Filled, nil
+	case "CANCELLED":
+		return order.Cancelled, nil
+	case "EXPIRED":
+		return order.Expired, nil
+	case "FAILED":
+		return order.Rejected, nil
+	default:
+		return order.UnknownStatus, fmt.Errorf("%w %v", errUnrecognisedStatusType, stat)
 	}
-	return err
 }
 
-func (c *CoinbasePro) authWsSubscibeReq(r *WebsocketSubscribe) error {
-	creds, err := c.GetCredentials(context.TODO())
-	if err != nil {
-		return err
+// stringToStandardType is a helper function that converts a Coinbase Pro side string to a standardised order.Type type
+func stringToStandardType(str string) (order.Type, error) {
+	switch str {
+	case "LIMIT_ORDER_TYPE":
+		return order.Limit, nil
+	case "MARKET_ORDER_TYPE":
+		return order.Market, nil
+	case "STOP_LIMIT_ORDER_TYPE":
+		return order.StopLimit, nil
+	default:
+		return order.UnknownType, fmt.Errorf("%w %v", order.ErrUnrecognisedOrderType, str)
 	}
-	r.Timestamp = strconv.FormatInt(time.Now().Unix(), 10)
-	message := r.Timestamp + http.MethodGet + "/users/self/verify"
-	hmac, err := crypto.GetHMAC(crypto.HashSHA256, []byte(message), []byte(creds.Secret))
-	if err != nil {
-		return err
+}
+
+// stringToStandardAsset is a helper function that converts a Coinbase Pro asset string to a standardised asset.Item type
+func stringToStandardAsset(str string) (asset.Item, error) {
+	switch str {
+	case "SPOT":
+		return asset.Spot, nil
+	case "FUTURE":
+		return asset.Futures, nil
+	default:
+		return asset.Empty, asset.ErrNotSupported
 	}
-	r.Signature = crypto.Base64Encode(hmac)
-	r.Key = creds.Key
-	r.Passphrase = creds.ClientID
-	return nil
 }
 
-// Unsubscribe sends a websocket message to stop receiving data from the channel
-func (c *CoinbasePro) Unsubscribe(subs subscription.List) error {
-	r := &WebsocketSubscribe{
-		Type:     "unsubscribe",
-		Channels: make([]any, 0, len(subs)),
+// strategyDecoder is a helper function that converts a Coinbase Pro time in force string to a few standardised bools
+func strategyDecoder(str string) (ioc, fok bool, err error) {
+	switch str {
+	case "IMMEDIATE_OR_CANCEL":
+		return true, false, nil
+	case "FILL_OR_KILL":
+		return false, true, nil
+	case "GOOD_UNTIL_CANCELLED", "GOOD_UNTIL_DATE_TIME":
+		return false, false, nil
+	default:
+		return false, false, fmt.Errorf("%w %v", errUnrecognisedStrategyType, str)
 	}
-	for _, s := range subs {
-		r.Channels = append(r.Channels, WsChannel{
-			Name:       s.Channel,
-			ProductIDs: s.Pairs.Strings(),
-		})
+}
+
+// Base64URLEncode is a helper function that does some tweaks to standard Base64 encoding, in a way which JWT requires
+func base64URLEncode(b []byte) string {
+	s := crypto.Base64Encode(b)
+	s = strings.Split(s, "=")[0]
+	s = strings.ReplaceAll(s, "+", "-")
+	s = strings.ReplaceAll(s, "/", "_")
+	return s
```

Comment:
```suggestion
	return base64.RawURLEncoding.EncodeToString(b)
```

Needs test coverage or removed because it doesn't seem to be used

Reply:
It's only used in my implementation of JWT, which is currently commented out due to not working for some indiscernible reason.